Introduce testing system with pgTAP

fljdin opened the pull request:
This PR adds a dedicated testing system to `db_migrator`, based on PGXS build infrastructure and pgTAP testing framework. A CONTRIBUTING.md is provided to explain how to run testing suite locally.

Add a no-operation (noop) plugin that rely on file_fdw and /dev/null. From my point of view, it will deliver more immediate feedback and quality than using other plugin as we did before. User tables and partitions tables must be attached to local datafiles (.dat) to test data insertion when `with_data` is enabled on `materialize_foreign_table()` calls.

Setup script executes `db_migrate_prepare()` function and populates staging tables with convenient datasets needed by pgTAP suite. Test ordering is managed by schedule files for `pg_regress` as setup must be executed before any others step, or even tables must exists before indexes creation.

laurenz commented:
An interesting initiative.

So far, I have used the regression test for ora_migrator to test db_migrator. But of course that requires that you have Oracle installed, and it can only test the features that Oracle supports (as we saw with the partial index patch, where the regression tests didn't catch that I messed up your code).

Let me know when you feel confident enough about the code that I should review it.

fljdin commented:
This PR is ready for review.

laurenz commented:
I expect to get to it this week.

laurenz commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+* [bug]: Something isn't working
+* [enhancement]: New feature or request
+* [help wanted]: Extra attention is needed
+* [problem]: Problem getting things to work, errors and the like 
```
That's pretty much all interesting categories.
Do we need such level of detail?

fljdin replied:
I was inspired by another projects' CONTRIBUTING.md
It's okay to remove it

laurenz replied:
I'm fine with `CONTRIBUTING.md` as such.
The project doesn't have the size that people are likely to get lost among the many pull requests and issues they want to work on...

laurenz commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,86 @@
+
+* PostgreSQL binary with PGXS build infrastructure (`devel` package)
+* a running PostgreSQL instance with valid credentials
+* pgTAP framework installed from [PGXN] with `pgxnclient`
```
I think you shouldn't require PGXN.
You can just as well install pgTAP without that.

laurenz commented on test/Makefile:
```diff
@@ -0,0 +1,31 @@
+	install -c -m 644 $(NOOP_DATA) $(INST_DIR)
+	sed -i "s;@testdata@;$(abspath $(TEST_DATA));" $(INST_DIR)/$(notdir $(NOOP_DATA))
+
+installcheck: installplugin
```
All your new files are missing a newline at the end...

laurenz reviewed (changes requested):
I just had a cursory look at the documentation and tried to run the new regression tests.

Unfortunately that failed:

```none
PGUSER=postgres PATH=/home/laurenz/pg/bin:$PATH make install

Nothing to be built.  Run "make install".
/usr/bin/mkdir -p '/home/laurenz/pg/share/postgresql/extension'
/usr/bin/mkdir -p '/home/laurenz/pg/share/postgresql/extension'
/usr/bin/mkdir -p '/home/laurenz/pg/share/doc/postgresql/extension'
/usr/bin/install -c -m 644 .//db_migrator.control '/home/laurenz/pg/share/postgresql/extension/'
/usr/bin/install -c -m 644 .//db_migrator--*.sql  '/home/laurenz/pg/share/postgresql/extension/'
/usr/bin/install -c -m 644 .//README.db_migrator '/home/laurenz/pg/share/doc/postgresql/extension/'
laurenz@~/db_migrator> PGUSER=postgres PATH=/home/laurenz/pg/bin:$PATH make installcheck
Install noop_migrator plugin
mkdir -p /home/laurenz/pg/share/postgresql/extension
install -c -m 644 test/noop_migrator--*.sql /home/laurenz/pg/share/postgresql/extension
sed -i "s;@testdata@;/home/laurenz/db_migrator/test/data;" /home/laurenz/pg/share/postgresql/extension/noop_migrator--*.sql
echo "# +++ regress install-check in  +++" && /home/laurenz/pg/lib/postgresql/pgxs/src/makefiles/../../src/test/regress/pg_regress --inputdir=./ --bindir='/home/laurenz/pg/bin'    --schedule test/schedules/setup.txt  --inputdir=test --outputdir=test --load-extension=pgtap --load-extension=file_fdw --load-extension=db_migrator --load-extension=noop_migrator --dbname=contrib_regression --schedule test/schedules/tests.txt 
# +++ regress install-check in  +++
# using postmaster on Unix socket, default port
ERROR:  extension "noop_migrator" is not available
DETAIL:  Could not open extension control file "/home/laurenz/pg/share/postgresql/extension/noop_migrator.control": No such file or directory.
HINT:  The extension must first be installed on the system where PostgreSQL is running.
# command failed: "/home/laurenz/pg/bin/psql" -X -q -c "CREATE EXTENSION IF NOT EXISTS \"noop_migrator\"" "contrib_regression"
Bail out!make: *** [/home/laurenz/pg/lib/postgresql/pgxs/src/makefiles/pgxs.mk:436: installcheck] Error 2
```

It seems like you forgot to install `noop_migrator.control`.

laurenz reviewed (changes requested):
The regression tests run now, but there are small differences in the output:

```none
diff -U3 /home/laurenz/db_migrator/test/expected/setup.out /home/laurenz/db_migrator/test/results/setup.out
--- /home/laurenz/db_migrator/test/expected/setup.out   2023-08-01 17:21:46.724618821 +0200
+++ /home/laurenz/db_migrator/test/results/setup.out    2023-08-01 17:23:55.178823818 +0200
@@ -1,4 +1,6 @@
 \i test/psql.sql
 \unset ECHO
 1..1
+NOTICE:  Creating staging schemas "fdw_stage" and "pgsql_stage" ...
+NOTICE:  Creating foreign metadata views in schema "fdw_stage" ...
 ok 1 - Should prepare staging schemas and catalog without errors
diff -U3 /home/laurenz/db_migrator/test/expected/tables.out /home/laurenz/db_migrator/test/results/tables.out
--- /home/laurenz/db_migrator/test/expected/tables.out  2023-08-01 17:21:46.724618821 +0200
+++ /home/laurenz/db_migrator/test/results/tables.out   2023-08-01 17:23:55.350825431 +0200
@@ -1,7 +1,13 @@
 \i test/psql.sql
 \unset ECHO
 1..6
+NOTICE:  Creating schemas ...
+NOTICE:  Creating sequences ...
+NOTICE:  Creating foreign tables ...
 ok 1 - Should create foreign tables without errors
+NOTICE:  Migrating table sch1.t1 ...
+NOTICE:  Migrating table sch1.t2 ...
+NOTICE:  Migrating table part1.pt1 ...
 ok 2 - Should migrate tables without errors
 ok 3 - Schema sch1 should have the correct tables
 ok 4 - Column t1.c1 should exist
diff -U3 /home/laurenz/db_migrator/test/expected/indexes.out /home/laurenz/db_migrator/test/results/indexes.out
--- /home/laurenz/db_migrator/test/expected/indexes.out 2023-08-01 17:21:46.723618812 +0200
+++ /home/laurenz/db_migrator/test/results/indexes.out  2023-08-01 17:23:55.526827082 +0200
@@ -1,6 +1,7 @@
 \i test/psql.sql
 \unset ECHO
 1..3
+NOTICE:  Creating indexes ...
 ok 1 - Should create user-defined indexes without errors
 ok 2 - Table sch1.t1 should have the correct indexes
 ok 3 - Index t1_idx2_unique should be unique
diff -U3 /home/laurenz/db_migrator/test/expected/constraints.out /home/laurenz/db_migrator/test/results/constraints.out
--- /home/laurenz/db_migrator/test/expected/constraints.out 2023-08-01 17:21:46.723618812 +0200
+++ /home/laurenz/db_migrator/test/results/constraints.out  2023-08-01 17:23:55.682828546 +0200
@@ -1,6 +1,10 @@
 \i test/psql.sql
 \unset ECHO
 1..7
+NOTICE:  Creating UNIQUE and PRIMARY KEY constraints ...
+NOTICE:  Creating FOREIGN KEY constraints ...
+NOTICE:  Creating CHECK constraints ...
+NOTICE:  Setting column default values ...
 ok 1 - Should create constraints without errors
 ok 2 - Table t1 should have a primary key
 ok 3 - Table t2 should not have a primary key
```

It seems that your installation suppresses notices. Perhaps you have `client_min_messages` configured differently?

laurenz committed:
Improve contributor documentation

Remove the paragraph that specifies which tags a new contributor
should work on.  We are not yet operating at the scale where this
would be necessary.  Let's add rules when we need them.

Other than that, I tried to improve the grammar, mostly by adding
articles.

This addresses #27#discussion_r1280738187

laurenz commented:
Would it be ok for you to develop in the [testing-with-pgtap](https://github.com/fljdin/db_migrator/tree/testing-with-pgtap) branch of this repository? Then we have fewer problems merging patches back and forth.

fljdin commented:
This PR is ready to be pushed into `cybertec-postgresql:testing-with-pgtap` branch.
I'll stop working on my fork, as you already suggested before. Good catch.

laurenz commented:
I notice that some features are untested:
- subpartitions
- views
- sequences
- functions
- triggers

I think it would be great if you added some tests for these as well.
